File: seed_variables.py
import os
import json
import re
from supabase import create_client, Client
from dotenv import load_dotenv

# Load env variables
load_dotenv()

SUPABASE_URL = os.getenv("SUPABASE_URL")
SUPABASE_KEY = os.getenv("SUPABASE_KEY") or os.getenv("SUPABASE_ANON_KEY")

if not SUPABASE_URL or not SUPABASE_KEY:
    print("Error: SUPABASE_URL or SUPABASE_KEY not found in environment.")
    exit(1)

supabase: Client = create_client(SUPABASE_URL, SUPABASE_KEY)

def parse_pipeline_service():
    file_path = "src/services/pipelineService.ts"
    if not os.path.exists(file_path):
        print(f"Error: {file_path} not found.")
        return []

    with open(file_path, "r", encoding="utf-8") as f:
        content = f.read()

    # Find the FIELD_MAP block
    match = re.search(r"export const FIELD_MAP = (\{.*?\});", content, re.DOTALL)
    if not match:
        print("Error: FIELD_MAP not found in pipelineService.ts")
        return []

    # This is a bit hacky because it's JS but we can try to turn it into JSON-ish
    # Or just extract fields manually with regex if the structure is consistent
    js_content = match.group(1)
    
    # Regex to capture individual entries
    # 'id': { label: '...', ... }
    entries = re.findall(r"'([\w\.]+)':\s*\{(.*?)\}", js_content, re.DOTALL)
    
    records = []
    for entry_id, body in entries:
        record = {"id": entry_id}
        
        # Extract properties
        label_m = re.search(r"label:\s*'(.*?)'", body)
        unit_m = re.search(r"unit:\s*'(.*?)'", body)
        desc_m = re.search(r"description:\s*'(.*?)'", body)
        module_m = re.search(r"module:\s*'(.*?)'", body)
        
        # Arrays
        keywords_m = re.search(r"keywords:\s*\[(.*?)\]", body)
        sources_m = re.search(r"sources:\s*\[(.*?)\]", body)
        
        if label_m: record["label"] = label_m.group(1)
        if unit_m: record["unit"] = unit_m.group(1)
        if desc_m: record["description"] = desc_m.group(1)
        if module_m: record["module"] = module_m.group(1)
        
        if keywords_m:
            record["keywords"] = [k.strip().strip("'").strip('"') for k in keywords_m.group(1).split(",") if k.strip()]
        if sources_m:
            record["sources"] = [s.strip().strip("'").strip('"') for s in sources_m.group(1).split(",") if s.strip()]
            
        records.append(record)
    
    return records

def seed_database(records):
    print(f"Seeding {len(records)} variables to Supabase...")
    
    # Chunking for bulk operations
    chunk_size = 50
    for i in range(0, len(records), chunk_size):
        chunk = records[i:i + chunk_size]
        try:
            response = supabase.table("variables").upsert(chunk).execute()
            print(f"Upserted chunk {i//chunk_size + 1}")
        except Exception as e:
            print(f"Error upserting chunk: {e}")

if __name__ == "__main__":
    variables = parse_pipeline_service()
    if variables:
        seed_database(variables)
    else:
        print("No variables found to seed.")

File: seed_variables.ts
import { createClient } from '@supabase/supabase-js';
import * as dotenv from 'dotenv';
import * as fs from 'fs';
import * as path from 'path';

dotenv.config();

const SUPABASE_URL = process.env.SUPABASE_URL;
const SUPABASE_KEY = process.env.SUPABASE_KEY || process.env.SUPABASE_ANON_KEY;

if (!SUPABASE_URL || !SUPABASE_KEY) {
  console.error("Error: SUPABASE_URL or SUPABASE_KEY not found in environment.");
  process.exit(1);
}

const supabase = createClient(SUPABASE_URL, SUPABASE_KEY);

function parsePipelineService() {
  const filePath = path.join(process.cwd(), 'src/services/pipelineService.ts');
  if (!fs.existsSync(filePath)) {
    console.error(`Error: ${filePath} not found.`);
    return [];
  }

  const content = fs.readFileSync(filePath, 'utf-8');

  // Find the FIELD_MAP block
  const match = content.match(/export const FIELD_MAP = (\{.*?\});/s);
  if (!match) {
    console.error("Error: FIELD_MAP not found in pipelineService.ts");
    return [];
  }

  const jsContent = match[1];
  
  // Regex to capture individual entries
  const entries = Array.from(jsContent.matchAll(/'([\w\.]+)':\s*\{(.*?)\}/gs));
  
  const records = entries.map(entry => {
    const [_, id, body] = entry;
    const record: any = { id };
    
    const labelM = body.match(/label:\s*'(.*?)'/);
    const unitM = body.match(/unit:\s*'(.*?)'/);
    const descM = body.match(/description:\s*'(.*?)'/);
    const moduleM = body.match(/module:\s*'(.*?)'/);
    
    const keywordsM = body.match(/keywords:\s*\[(.*?)\]/s);
    const sourcesM = body.match(/sources:\s*\[(.*?)\]/s);
    
    if (labelM) record.label = labelM[1];
    if (unitM) record.unit = unitM[1];
    if (descM) record.description = descM[1];
    if (moduleM) record.module = moduleM[1];
    
    if (keywordsM) {
      record.keywords = keywordsM[1].split(',')
        .map(k => k.trim().replace(/^['"]|['"]$/g, ''))
        .filter(k => k);
    }
    if (sourcesM) {
      record.sources = sourcesM[1].split(',')
        .map(s => s.trim().replace(/^['"]|['"]$/g, ''))
        .filter(s => s);
    }
    
    return record;
  });
  
  return records;
}

async function seedDatabase(records: any[]) {
  console.log(`Seeding ${records.length} variables to Supabase...`);
  
  const chunkSize = 50;
  for (let i = 0; i < records.length; i += chunkSize) {
    const chunk = records.slice(i, i + chunkSize);
    const { error } = await supabase.from('variables').upsert(chunk);
    if (error) {
      console.error(`Error upserting chunk ${i / chunkSize + 1}:`, error);
    } else {
      console.log(`Upserted chunk ${i / chunkSize + 1}`);
    }
  }
}

async function main() {
  const variables = parsePipelineService();
  if (variables.length > 0) {
    await seedDatabase(variables);
  } else {
    console.log("No variables found to seed.");
  }
}

main().catch(console.error);
